videostore/indexer.go

```
@@ -0,0 +1,388 @@
+func (ix *Indexer) scanAndIndex() error {
+	// Get disk files
+	diskFiles, err := ix.getDiskFiles()
+	if err != nil {
+		ix.logger.Warnw("failed to list disk files, skipping scan cycle", "path", ix.storagePath, "error", err)
+		return nil
+	}
+
+	// Get indexed files
+	indexedFiles, err := ix.getIndexedFiles()
+	if err != nil {
+		return fmt.Errorf("failed to get indexed files from DB: %w", err)
+	}
+
+	// Diff disk files and indexed files
+	for path, info := range diskFiles {
+		if _, exists := indexedFiles[path]; !exists {
+			ix.logger.Debugf("found new segment to index: %s", path)
+			if err := ix.indexNewFile(path, info); err != nil {
+				ix.logger.Errorw("failed to index new file", "path", path, "error", err)
+			}
+		}
+	}
+
+	return nil
+}
+
+// getIndexedFiles retrieves the set of file paths currently in the index.
+func (ix *Indexer) getIndexedFiles() (map[string]struct{}, error) {
+	query := "SELECT file_path FROM " + segmentsTableName + ";"
+	rows, err := ix.db.Query(query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	files := make(map[string]struct{})
+	for rows.Next() {
+		var filePath string
+		if err := rows.Scan(&filePath); err != nil {
+			return nil, err
+		}
+		files[filePath] = struct{}{}
+	}
+	return files, rows.Err()
+}
+
+// getDiskFiles retrieves the set of .mp4 files currently in the storage directory.
+// It interacts with the filesystem only and does not require the indexer mutex.
+func (ix *Indexer) getDiskFiles() (map[string]os.FileInfo, error) {
+	files := make(map[string]os.FileInfo)
+	entries, err := os.ReadDir(ix.storagePath)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, entry := range entries {
+		if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".mp4") {
+			info, err := entry.Info()
+			if err != nil {
+				ix.logger.Warnw("failed to get FileInfo for disk file", "name", entry.Name(), "error", err)
+				continue
```

lets not continue here, lets return an error. Unless we know of a specific reason why this would happen we should bias towards not omitting information on error.

---

Changed to err return